bundle/jsky.app.ot/src/main/scala/jsky/app/ot/tpe/feat/TpeGhostIFUFeature.scala

```diff
@@ -0,0 +1,505 @@
+
+
+  override def getKey: JOption[Component] = {
+    val pan = new JPanel(new GridBagLayout)
+
+    val ifu1Label: JLabel = new JLabel("IFU1", new ProbeRangeIcon(Array(TpeGhostIfuFeature.NorthEast)), SwingConstants.LEFT)
+    pan.setForeground(Color.black)
+    val gbc1: GridBagConstraints = new GridBagConstraints(0, 0, 1, 1, 0,
+      0, java.awt.GridBagConstraints.CENTER, java.awt.GridBagConstraints.HORIZONTAL,
+      new Insets(0, 0, 0, 0), 0, 0)
+
+
+    val ifu2Label: JLabel = new JLabel("IFU2", new ProbeRangeIcon(Array(TpeGhostIfuFeature.SouthEast)), SwingConstants.LEFT)
+    ifu2Label.setForeground(Color.black)
+    val gbc2: GridBagConstraints = new GridBagConstraints(1, 0, 1, 1, 0,
+      0, java.awt.GridBagConstraints.CENTER, java.awt.GridBagConstraints.HORIZONTAL,
+      new Insets(0, 0, 0, 0), 0, 0)
+
+    val bothLabel: JLabel = new JLabel("Both", new ProbeRangeIcon(Array(TpeGhostIfuFeature.NorthEast, TpeGhostIfuFeature.SouthEast)), SwingConstants.LEFT)
+    bothLabel.setForeground(Color.black)
+    val gbcBoth: GridBagConstraints = new GridBagConstraints(2, 0, 1, 1, 0,
+      0, java.awt.GridBagConstraints.CENTER, java.awt.GridBagConstraints.HORIZONTAL,
+      new Insets(0, 0, 0, 0), 0, 0)
+
+
+    pan.add(ifu1Label, gbc1)
+    pan.add(ifu2Label, gbc2)
+    pan.add(bothLabel, gbcBoth)
+
+    new JSome[Component](pan)
+  }
+
+  override def getMessages: JOption[java.util.Collection[TpeMessage]] = {
+    if (!isEmpty) JNone.instance()
+    else new JSome(Collections.singletonList(TpeGhostIfuFeature.Warning))
+  }
+}
+
+
+object TpeGhostIfuFeature {
+  /**
+   * Calculates the current patrol field for an IFU as specified by its default patrol field.
+   */
+  private def ifuPatrolField(ctx: ObsContext, patrol: Rectangle2D.Double): Area = {
+    val rot = if (ctx.getIssPort == IssPort.SIDE_LOOKING) Angle.fromDegrees(90) else Angle.fromDegrees(0)
+    val t: Double = ctx.getPositionAngle.toRadians + rot.toRadians
+
+    val rects = ctx.getSciencePositions.asScala.toSet.map { pos: Offset =>
+      val fov = new Area(GhostScienceAreaGeometry.Ellipse)
+      val offsetPatrol = new Area(patrol)
+
+      val p = pos.p.toArcsecs.getMagnitude
+      val q = pos.q.toArcsecs.getMagnitude
+      val xform = new AffineTransform
+      if (t != 0.0) xform.rotate(-t)
+      xform.translate(-p, -q)
+
+      fov.transform(xform)
+      offsetPatrol.transform(xform)
+      offsetPatrol.intersect(fov)
+      offsetPatrol
+    }
+
+    rects.foldLeft(rects.head) { (acc, curr) =>
+      acc.intersect(curr)
+      acc }
+  }
+
+  // The patrol fields for IFU1 and IFU2 relative to the base position.
+  private val ifu1PatrolField = new Rectangle2D.Double(-222, -222, 222 + 3.28, 444)
+  private val ifu2PatrolField = new Rectangle2D.Double(-3.28, -222, 222 + 3.28, 444)
+  private def ifu1Arc(ctx: ObsContext): Area = ifuPatrolField(ctx, ifu1PatrolField)
+  private def ifu2Arc(ctx: ObsContext): Area = ifuPatrolField(ctx, ifu2PatrolField)
```

Comment:
We're mixing model elements with drawing code in this class, like we do most everywhere else.  I'm not sure it's worth untangling right now but ideally this would have gone down in the model like we do for the GMOS OIWFS and detector.

Reply:
Good point.